proposals/base-relative-assets.md

```diff
@@ -0,0 +1,132 @@
+
+- For link tags it is done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L300).
+- For hoisted scripts it is [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L301)
+- For links added to astro-islands it [is here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L352).
+
+We should generalize this logic so that there's one function that gets called to resolve the URL given the `linkStyle` option, as well as the base and subpath. That function will use the `linkStyle` config to determine which algorithm to use.
+
+For SSR the same things are done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/app/index.ts#L63) for example. Again, this should be consolidated so we're not doing the same sort of logic in many different places, creating areas for bugs. Instead SSG and SSR should use shared logic.
+
+## 404.astro
+
+If using `linkStyle: 'relative'` a 404.astro page __will not be allowed__. This is because it is impossible to make these links be relative in SSG mode. That's because a 404.astro might be used by paths such as:
+
+- /foo/
+- /foo/bar/
+- /foo/bar/baz/
+
+If we statically generate relative links then these will be incorrect. So we will throw an error if using a 404.astro page with `'relative'` and suggest using `'absolute'` instead.
+
+## Imported images
+
+Images that are imported will *not* be relative. That means this:
+
+```astro
+---
+import penguinUrl from '../assets/penguin.png';
+---
+
+<img src={penguinUrl} />
+```
+
+It might be *possible* to make this work through a Vite plugin but would be a bit of work and might be error prone. 
+
+# Testing strategy
+
+Given that resolving URLs will be done in a few places, it makes sense to create a module within `src/core` for this algorithm. This module will be unit testable for all of the scenarios. We should probably also move the runtime resolver (which is used internally by the runtime for producing `<astro-island>` URLs) since it is shared logic between SSG and SSR that is related.
+
+Additionally some fixture tests will be included for the matrix of scenarios; SSG and SSR mode, `absolute` and `relative` options, as well as the types of assets; stylesheets, hoisted scripts, and astro islands.
+
+# Drawbacks
+
+- This adds some complexity to the codebase in the usual way that adding a config option does.
+  - However this is also an opportunity to clean up the code and have just 1 algorithm for determing how a URL gets resolved that we create.
+- This will not fix the problem of `<img>`s you create. this is especially important in Markdown files, where manually adding relative links is required. So people might find this solution inadequate since it doesn't fix all of your problems.
```

Comment:
@delucis since you know a lot about custom md syntax, is it possible to have a syntax that would be a little like a directive is in Astro where the user can say "this should be relative" for an image?